### PR TITLE
swarm-derive: Allow for templated behaviours

### LINKED ISSUE #2879: `#[derive(NetworkBehaviour)]` fails on structs with generics
## Summary

It looks like there is a regression in libp2p-0.48.0: `#[derive(NetworkBehaviour)]` fails to generate code if the original struct has generic parameters.

## Expected behaviour

`#[derive(NetworkBehaviour)]` generates the implementation if the struct has generic parameters.

## Actual behaviour

Compilation fails with the error:
```
error: proc-macro derive produced unparseable tokens
```

Please see the minimal example: [`main.rs`](https://github.com/dmitry-markin/libp2p-behaviour/blob/916d8b8522901f996efa857c11b924a70c12a9a8/src/main.rs)

(In this case the compilation would fail anyway, because `Option<T>` does not implement `NetworkBehaviour`, but the compilation fails earlier with the same error as in the real-life example in Substrate, where the templatized type actually implements `NetworkBehaviour`.)

## Version
- libp2p version: 0.48.0

## Additional information

It also looks like there is another, may be connected, bug: `#[derive(NetworkBehaviour)]` fails to generate `<STRUCT_NAME>Event` if no `#[behaviour(out_event = "...")]` provided (even without generics):
```
error[E0433]: failed to resolve: use of undeclared type `NetworkBehaviour`
```

## Would you like to work on fixing this bug?

No, unfortunately I don't understand what's happening.

CC @mxinden

### PR DESCRIPTION
**Pull request by thomaseizinger**
# Description

This patch fixes an issue where we couldn't use type parameters on a behaviour with a custom out event that had different type parameters.

## Links to any relevant issues

Fixes #2879.

## Change checklist

- [x] I have performed a self-review of my own code
~- [ ] I have made corresponding changes to the documentation~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm-derive/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,10 @@
+# 0.30.1
+
+- Fix an issue where the derive would generate bad code if the type parameters between the behaviour and a custom
+  out event differed. See [PR XXXX].
+
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (thomaseizinger) [inline swarm-derive/CHANGELOG.md]:
```suggestion
- Fix an issue where the derive would generate bad code if the type parameters between the behaviour and a custom
  out event differed. See [PR 2907].

[PR 2907]: https://github.com/libp2p/rust-libp2p/pull/2907
```

**Review by mxinden — APPROVED** [review]
Thanks for debugging and patching @thomaseizinger. Thanks @dmitry-markin for confirming.

@dmitry-markin let me know in case you need this released rather sooner than later.

**Comment by mxinden** [comment]
Testground failures are due to https://github.com/libp2p/test-plans/pull/41. Merging here.

**Comment by dmitry-markin** [comment]
> @dmitry-markin let me know in case you need this released rather sooner than later.

@mxinden it would be great to have this released to merge https://github.com/paritytech/substrate/pull/12256 and be able to finally use https://github.com/libp2p/rust-libp2p/pull/2846. But this is not super-urgent.

**Comment by mxinden** [comment]
@dmitry-markin you can follow the progress on the next release here: https://github.com/libp2p/rust-libp2p/pull/2931

**Pull request merged**